common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,438 @@
+    [&](const Model::BrushNode* brush) { updateBrush(brush); },
+    [](Model::PatchNode*) {}));
+}
+
+void EntityDecalRenderer::removeNode(Model::Node* node)
+{
+  node->accept(kdl::overload(
+    [](Model::WorldNode*) {},
+    [](Model::LayerNode*) {},
+    [](Model::GroupNode*) {},
+    [&](const Model::EntityNode* entity) { removeEntity(entity); },
+    [&](const Model::BrushNode* brush) { removeBrush(brush); },
+    [](Model::PatchNode*) {}));
+}
+
+std::optional<Assets::DecalSpecification> EntityDecalRenderer::getDecalSpecification(
+  const Model::EntityNode* entityNode)
+{
+  const auto decalSpec = entityNode->entity().decalSpecification();
+  return decalSpec.textureName.empty() ? std::nullopt : std::make_optional(decalSpec);
+}
+
+void EntityDecalRenderer::updateEntity(const Model::EntityNode* entityNode)
+{
+  // if the entity isn't visible, don't create decal geometry for it
+  const auto& editorContext = kdl::mem_lock(m_document)->editorContext();
+  if (!editorContext.visible(entityNode))
+  {
+    return;
+  }
+
+  // check if the entity has a decal specification
+  const auto spec = getDecalSpecification(entityNode);
+
+  // see if we are tracking this entity
+  const auto ent = m_entities.find(entityNode);
+  const auto isTracking = ent != std::end(m_entities);
+  if (isTracking && spec.has_value())
+  {
+    // entity is being tracked and has a decal specification, invalidate it
+    invalidateDecalData(ent->second);
+  }
+  else if (isTracking)
+  {
+    // entity is being tracked but no longer has a decal specification
+    removeEntity(entityNode);
+  }
+  else if (spec.has_value())
+  {
+    // entity is not being tracked and has a decal specification, start tracking it
+    m_entities.insert({entityNode, EntityDecalData{}});
+  }
+}
+
+void EntityDecalRenderer::removeEntity(const Model::EntityNode* entityNode)
+{
+  if (const auto it = m_entities.find(entityNode); it != std::end(m_entities))
+  {
+    // make sure the entity data is cleaned up
+    invalidateDecalData(it->second);
+    m_entities.erase(it);
+  }
+}
+
+void EntityDecalRenderer::updateBrush(const Model::BrushNode* brushNode)
+{
+  // invalidate any entities that intersect this brush or are tracking this brush
+  for (auto& [ent, data] : m_entities)
+  {
+    // skip entities that are going to be recomputed anyway
+    if (!data.validated)
+    {
+      continue;
+    }
+
+    // if the brush is not visible, then it doesn't (currently) intersect
+    const auto& editorContext = kdl::mem_lock(m_document)->editorContext();
+    const auto intersects =
+      editorContext.visible(brushNode) && brushNode->intersects(ent);
+    const auto tracked = std::find(data.brushes.begin(), data.brushes.end(), brushNode)
+                         != data.brushes.end();
+
+    // if this brush is tracked by this entity or intersects, we'll need to
+    // recalculate the geometry
+    if (intersects || tracked)
```

Comment:
Same here, if the brush has become invisible, it needs to be removed, doesn't it?

Reply:
When the decal is re-validated, the list of intersecting (visible) brushes will be recalculated, so I think for this case it will work fine. (if a brush is made invisible, then `intersects` will be `false` and `tracked` will be `true`, which will invalidate the decal)